core/drivers/hi16xx_uart.c

```diff
@@ -88,47 +86,29 @@ void hi16xx_uart_init(vaddr_t base, uint32_t uart_clk, uint32_t baud_rate)
 {
 	uint16_t freq_div = uart_clk / (16 * baud_rate);
 
-	/* Clear and enable FIFO */
-	write32(UART_FCR_FIFO_EN | UART_FCR_RX_FIFO_RST | UART_FCR_TX_FIFO_RST,
-		base + UART_FCR);
-	dsb();
+	/* Enable (and clear) FIFOs */
+	write32(UART_FCR_FIFO_EN, base + UART_FCR);
 
 	/* Enable access to _DLL and _DLH */
 	write32(UART_LCR_DLAB, base + UART_LCR);
-	dsb();
 
 	/* Calculate and set UART_DLL */
 	write32(freq_div & 0xFF, base + UART_DLL);
-	dsb();
 
 	/* Calculate and set UART_DLH */
 	write32((freq_div >> 8) & 0xFF, base + UART_DLH);
-	dsb();
 
-	/*
-	 * Clear _DLL and _DLH access bit, set data size (8 bits), parity etc.
-	 */
+	/* Clear _DLL/_DLH access bit, set data size (8 bits), parity etc. */
 	write32(UART_LCR_DLS8, base + UART_LCR);
-	dsb();
 
 	/* Disable interrupt mode */
 	write32(0, base + UART_IEL);
-	dsb();
```

Comment:
I think the device mapping will preserve ordering, at least within a small region. This particular device is less than 1k, so we may be OK, and the read will happen after the writes before it, so it is possible things will be OK without any barriers.

We do have other devices that are larger than 1k. It's only a requirement, and a lot of SOC architectures preserve the ordering in larger areas than that (I do know that some Qualcomm devices treat the 1k strictly, and will reorder on operations across larger addresses), I'm not sure what this device does, but it is probably worth asking if maybe we can find out.